python/pyspark/testing/utils.py

```diff
@@ -209,3 +234,79 @@ def check_error(
         self.assertEqual(
             expected, actual, f"Expected message parameters was '{expected}', got '{actual}'"
         )
+
+
+def assertDataFrameEqual(df: DataFrame, expected: DataFrame, ignore_row_order: bool = True):
+    if df is None and expected is None:
+        return True
+    elif df is None or expected is None:
+        return False
+
+    def compare_rows(r1: Row, r2: Row):
+        def compare_vals(val1, val2):
+            if isinstance(val1, list) and isinstance(val2, list):
+                return all([compare_vals(x, y) for x, y in list(zip(val1, val2))])
```

Comment:
nit: Shall we avoid surrounding with `[]` to leverage the short-circuit path?

```py
all(compare_vals(x, y) for x, y in zip(val1, val2))
```

also `list` is redundant.